### PR TITLE
fix(packagist): De-minify fields

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- De-minify fields at the schema level
- [Code ported from Composer-related library](https://github.com/composer/metadata-minifier/blob/c549d23829536f0d0e984aaabbf02af91f443207/src/MetadataMinifier.php#L17)

## Context

- Ref: #18715

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/packagist/schema.ts [hunk 1]
```diff
@@ -14,14 +14,54 @@ export const ComposerRelease = z.object({
...
+
+// Ported from: https://github.com/composer/metadata-minifier/blob/main/src/MetadataMinifier.php#L17
+function expandReleases(releases: ComposerRelease[]): ComposerRelease[] {
+  let prevReleaseVals: MinifiedRelease | null = null;
+
+  for (const composerRelease of releases) {
+    if (!prevReleaseVals) {
+      prevReleaseVals = {
+        homepage: composerRelease.homepage,
+        source: composerRelease.source,
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:38]: I'm not sure these should be included for each _release_ as they rarely change? I would prefer it's like this:
- Define the homepage and sourceUrl for the package based on the latest release
- Only add homepage and sourceUrl to individual releases unless they are different

File: lib/modules/datasource/packagist/schema.ts [hunk 2]
```diff
@@ -14,14 +14,54 @@ export const ComposerRelease = z.object({
     .transform((x) => x.url)
     .nullable()
     .catch(null),
+  require: z
+    .object({
+      php: z.string(),
+    })
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:20]: Based on their docs it sounds like either `require` or `php` could possibly be a string called `__unset`. Will this still gracefully ignore that if found?

File: lib/modules/datasource/packagist/schema.ts [hunk 3]
```diff
@@ -14,14 +14,54 @@ export const ComposerRelease = z.object({
...
+
+    for (const k of Object.keys(prevReleaseVals)) {
+      const key: keyof MinifiedRelease = k as keyof MinifiedRelease;
+      const prevReleaseVal = prevReleaseVals[key];
+      const currentReleaseVal = composerRelease[key];
+      if (currentReleaseVal !== null) {
+        prevReleaseVals[key] = currentReleaseVal;
+      } else if (prevReleaseVal !== null) {
+        composerRelease[key] = prevReleaseVal;
+      }
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:52]: Not sure this is the same as the `__unset` logic in composer?

File: lib/modules/datasource/packagist/schema.ts [hunk 4]
```diff
@@ -65,8 +105,14 @@ export function parsePackagesResponses(
...
-        dep.releaseTimestamp = composerRelease.time;
+      const { time, require } = composerRelease;
+
+      if (time) {
+        dep.releaseTimestamp = time;
+      }
+
+      if (require) {
+        dep.constraints = { php: [require] };
       }
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:116]: This bit is a bit confusing. Within the json it's require.php

File: lib/modules/datasource/packagist/schema.ts (hunk 3 above)
Comment (herndlm) [reply lib/modules/datasource/packagist/schema.ts:52]: I pulled https://3v4l.org/IjoG6L out of their test suite, maybe this helps you verify the logic. It prints a minified and an expanded array with the exact same expand logic as in the lib

File: lib/modules/datasource/packagist/schema.ts [hunk 5]
```diff
@@ -1,27 +1,92 @@
...
+  releases: ComposerReleaseMinified[]
+): ComposerRelease[] {
+  const prevVals: MinifiableReleaseValues = {
+    require: null,
+  };
+
+  for (const release of releases) {
+    for (const k of Object.keys(prevVals)) {
+      const key: keyof MinifiableReleaseValues =
+        k as keyof MinifiableReleaseValues;
```
Comment (viceice) [inline lib/modules/datasource/packagist/schema.ts:61]:
```suggestion
    for (const key of Object.keys(prevVals) as (keyof MinifiableReleaseValues)[]) {
```

simplify?

File: lib/modules/datasource/packagist/schema.ts [hunk 6]
```diff
@@ -68,6 +112,10 @@ export function parsePackagesResponses(
         dep.releaseTimestamp = composerRelease.time;
       }
 
+      if (composerRelease.require?.php) {
+        dep.constraints = { php: [composerRelease.require.php] };
+      }
+
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:118]: Maybe separate this?

File: lib/modules/datasource/packagist/schema.ts [hunk 7]
```diff
@@ -17,6 +54,12 @@ export const ComposerRelease = z
           .nullable()
           .catch(null),
         time: z.string().nullable().catch(null),
+        require: z
+          .object({
+            php: z.string(),
+          })
+          .nullable()
+          .catch(null),
```
Comment (rarkins) [inline lib/modules/datasource/packagist/schema.ts:62]:
```suggestion
```

I suggest we separate out the `require.php` gathering into a future PR

Comment (rarkins) [reply lib/modules/datasource/packagist/schema.ts:62]: Although if you think that's too painful, I don't mind keeping it

Comment (zharinov) [reply lib/modules/datasource/packagist/schema.ts:62]: Haha, no, I just missed it

**Review by rarkins — APPROVED** [review]

**Pull request merged**